sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureBlobFileAttributes.java

```diff
@@ -43,14 +45,24 @@
 
     private final BlobProperties properties;
     private final AzureResource resource;
+    private final boolean isVirtualDirectory;
 
     AzureBlobFileAttributes(Path path) throws IOException {
+        this.resource = new AzureResource(path);
+        BlobProperties props = null;
         try {
-            this.resource =  new AzureResource(path);
-            this.properties = resource.getBlobClient().getProperties();
+            props = resource.getBlobClient().getProperties();
         } catch (BlobStorageException e) {
-            throw LoggingUtility.logError(logger, new IOException("Path: " + path.toString(), e));
+            if (e.getStatusCode() == 404 && this.resource.checkDirectoryExists()) {
```

Comment:
It seems that `this.resource.checkDirectoryExists()` internally calls `this.getBlobClient().getProperties()` in the `checkDirStatus()`. Which means have one more extra get properties call. We should work a bit on this to make sure this only calls blob listing as a fallback.

Reply:
Yea it's not perfectly optimal, I agree. I think in general checkDirectoryStatus() (and consequently checkDirectoryExists()) may be trying to do too much at once and should potentially be refactored. That said, given that we've had more requests for functionality than performance, I've been trying to avoid premature optimization. Especially in this case, I don't know how common it is that people will be trying to get file attributes on a nonexistent file/virtual directory. Thoughts?

Reply:
If the penalty was only runtime performance I would agree. However, in this case the penalty is extra $ spent by customer unnecessarily.

Reply:
Fair enough. I think I can refactor the listing part of that method into its own method and just use that here